bolt/bbolt.go

```diff
@@ -118,32 +118,49 @@ func (db *DB) List(bucket []byte) ([]*database.Entry, error) {
 	return entries, err
 }
 
-// LoadOrStore stores a id/value pair in the table if the id is not yet set.
-func (db *DB) LoadOrStore(bucket, key, value []byte) ([]byte, bool, error) {
-	var (
-		ret   []byte
-		found bool
-	)
-	err := db.db.Update(func(boltTx *bolt.Tx) (err error) {
-		b, err := db.getBucket(boltTx, bucket)
-		if err != nil {
-			return err
+// CmpAndSwap modifies the value at the given bucket and key (to newValue)
+// only if the existing (current) value matches oldValue.
+func (db *DB) CmpAndSwap(bucket, key, newValue, oldValue []byte) ([]byte, bool, error) {
+	boltTx, err := db.db.Begin(true)
+	if err != nil {
+		return nil, false, errors.Wrap(err, "error creating Bolt transaction")
+	}
+
+	boltBucket := boltTx.Bucket(bucket)
+	if boltBucket == nil {
+		return nil, false, errors.Errorf("failed to get bucket %s", bucket)
+	}
+
+	val, swapped, err := cmpAndSwap(boltBucket, key, newValue, oldValue)
+	switch {
+	case err != nil:
+		if err := boltTx.Rollback(); err != nil {
+			return nil, false, errors.Wrapf(err, "failed to execute CmpAndSwap transaction on %s/%s and failed to rollback transaction", bucket, key)
 		}
-		ret = b.Get(key)
-		switch {
-		case ret == nil:
-			// Not Found, so try to set.
-			if err = b.Put(key, value); err != nil {
-				return errors.WithStack(err)
-			}
-			return nil
-		default:
-			// Found
-			found = true
-			return nil
+		return nil, false, err
+	case swapped:
+		if err := boltTx.Commit(); err != nil {
+			return nil, false, errors.Wrapf(err, "failed to commit badger transaction")
 		}
-	})
-	return ret, found, err
+		return val, swapped, nil
+	default:
+		if err := boltTx.Rollback(); err != nil {
+			return nil, false, errors.Wrapf(err, "failed to rollback read-only CmpAndSwap transaction on %s/%s", bucket, key)
+		}
+		return val, swapped, err
+	}
+}
+
+func cmpAndSwap(boltBucket *bolt.Bucket, key, newValue, oldValue []byte) ([]byte, bool, error) {
+	current := boltBucket.Get(key)
+	if !bytes.Equal(current, oldValue) {
+		return current, false, nil
```

Comment:
You might need to return a copy of `current`. According to bbolt docs:
```
func (b *Bucket) Get(key []byte) []byte

Get retrieves the value for a key in the bucket. Returns a nil value if the key
does not exist or if the key is a nested bucket. The returned value is only
valid for the life of the transaction.
```
Check if this is properly done in the other implementations

Reply:
good catch.